plugins/llvm/llvm_binary.hpp

```diff
@@ -108,52 +108,24 @@ using namespace llvm;
 using namespace llvm::object;
 
 struct segment {
-    template <typename T>
-    segment(const Elf_Phdr_Impl<T>& hdr, int pos)
-        : offset_(hdr.p_offset)
-        , addr_(hdr.p_vaddr)
-        , size_(hdr.p_filesz)
-        , is_readable_(hdr.p_flags & ELF::PF_R)
-        , is_writable_(hdr.p_flags & ELF::PF_W)
-        , is_executable_(hdr.p_flags & ELF::PF_X) {
-        std::ostringstream oss;
-        oss << std::setfill('0') << std::setw(2) << pos ;
-        name_ = oss.str();
-    }
-
-    segment(const MachO::segment_command &s) {
-        init_macho_segment(s);
-    }
-
-    segment(const MachO::segment_command_64 &s) {
-        init_macho_segment(s);
-    }
-
-    segment(const pe32_header &hdr, const coff_section &s)
-        : name_(s.Name)
-        , offset_(static_cast<uint32_t>(s.PointerToRawData))
-        , addr_(static_cast<uint32_t>(s.VirtualAddress + hdr.ImageBase))
-        , size_(static_cast<uint32_t>(s.SizeOfRawData))
-        , is_readable_(static_cast<uint32_t>(s.Characteristics) &
-                       COFF::IMAGE_SCN_MEM_READ)
-        , is_writable_(static_cast<uint32_t>(s.Characteristics) &
-		       COFF::IMAGE_SCN_MEM_WRITE)
-	, is_executable_(static_cast<uint32_t>(s.Characteristics) &
-			 COFF::IMAGE_SCN_MEM_EXECUTE)
+    segment(std::string name, uint64_t offset, uint64_t addr, 
+	    uint64_t size, bool r, bool w, bool x)
+	: name_(name)
+	, offset_(offset)
+	, addr_(addr)
+	, size_(size)
+	, is_readable_(r)
+	, is_writable_(w)
+	, is_executable_(x)
     {}
 
-    segment(const pe32plus_header &hdr, const coff_section &s)
-        : name_(s.Name)
-        , offset_(static_cast<uint64_t>(s.PointerToRawData))
-	, addr_(static_cast<uint64_t>(s.VirtualAddress + hdr.ImageBase))
-        , size_(static_cast<uint64_t>(s.SizeOfRawData))
-        , is_readable_(static_cast<uint64_t>(s.Characteristics) &
-                       COFF::IMAGE_SCN_MEM_READ)
-        , is_writable_(static_cast<uint64_t>(s.Characteristics) &
-                       COFF::IMAGE_SCN_MEM_WRITE)
-        , is_executable_(static_cast<uint64_t>(s.Characteristics) &
-                         COFF::IMAGE_SCN_MEM_EXECUTE)
-    {}
+    template <typename S>
```

Comment:
Remove this code from the constructor. This is a code that is specific to a particular format. Nothing format specific should be in the `segment` class.